### PR TITLE
Fix issue#1

### LINKED ISSUE #1: Implement the Card array generation logic
# Description
We need to implement the card generation logic in the deck.js file. This is a crucial part of our application as it will be responsible for creating the deck of cards that will be used in the game.

Please refer to [deck.js](https://github.com/shivansh-bhatnagar18/multiplayer-uno/blob/master/backend/uno-game-engine/deck.js)

## Task
The array `deck` should be populated with the cards as per the rules of the game, in the structure mentioned in the comment.

### PR DESCRIPTION
**Pull request by Abhishek-Punhani**
backend/deck.js
Fixed the function to initialize the deck according to the requirements and added valid checks to check the composition of deck

Fixes: #1

### CODE REVIEW

File: backend/tests/deck.test.ts [hunk 1]
```diff
@@ -1,7 +1,11 @@
-import { makeCard } from '../uno-game-engine/deck';
+import getShuffledCardDeck, { makeCard } from '../uno-game-engine/deck';
 describe('testing deck.ts', () => {
     test('makeCard', () => {
         const card = makeCard('number', 'blue', '3');
         expect(card.color).toBe('blue');
     });
+    test('getShuffledDeck', () => {
+        const deck = getShuffledCardDeck();
+        console.log(deck);
```
Comment (kuv2707) [inline backend/tests/deck.test.ts:9]: The purpose of tests is not to just console log things. That might have been required during development, but here, you should test whether the card deck correctly creates all the required cards. For example, you should expect there to be 25 cards per color, and 19 of them should be number cards, etc..

File: backend/uno-game-engine/deck.ts [hunk 2]
```diff
@@ -14,39 +15,57 @@ const values = [
...
-    const deck = [];
-    // todo: Implement the card generation logic
-    // dummy code:
-    // deck.push(makeCard('special', 'wild', 'wild'))
-    // deck.push(makeCard('number', 'red', '0'))
+This function returns a shuffled deck of 108 UNO cards. Each card is an object with a color and a value.
+The function makeCard is used to make the card objects.
+@returns {Array} deck - An array of 108 UNO cards.
+*/
+export default function getShuffledCardDeck(): Array<UNOCard> {
```
Comment (kuv2707) [inline backend/uno-game-engine/deck.ts:45]: Can you remove `default` from here?

File: backend/uno-game-engine/deck.ts [hunk 3]
```diff
@@ -14,39 +15,57 @@ const values = [
...
+    const deck: Array<UNOCard> = [];
+
+    for (let col of colors) {
+        for (let val of values) {
+            if (!isNaN(Number(val))) {
+                for (let i = 0; i < 2; i++) {
+                    deck.push(makeCard('number', col, val));
+                    if (val === '0') break;
+                }
+            } else {
```
Comment (kuv2707) [inline backend/uno-game-engine/deck.ts:55]: Great job on the algorithm!

**Review by kuv2707 — COMMENTED** [review]
Suggested some changes. Focus on improving the tests. Look up jest testing library.

Also the commit title of a preparatory commit doesn't literally have to say so. The commit message should be the same as any other commit, whose format is mentioned in the docs. (just that the Fixes clause is omitted, as that commit doesn't actually fix the issue)

Side note: You can use github's request review feature once you are done with the changes.
Additionally, the PR description can be a bit more descriptive.

File: backend/tests/deck.test.ts (hunk 1 above)
Comment (Abhishek-Punhani) [reply backend/tests/deck.test.ts:9]: Yes i have checked them all like each color cards, special cards , and length of deck to be 108

Comment (kuv2707) [reply backend/tests/deck.test.ts:9]: What I meant was for you to write a test which counts the number of red green etc and assert that they are 25
Then count the no. of wildcards and assert that there are 8, etc.

Writing tests ensures that future changes do not break existing functionality.

**Comment by kuv2707** [comment]
@Abhishek-Punhani Since this is a high priority issue, I'll have to go with #30 if the requested changes are not made by today evening.

**Pull request closed without merging**